rust/template/cmd_parser/lib.rs

```diff
@@ -39,88 +63,290 @@ where
+                        Ok(0) => {
+                            return Ok(());
+                        }
+                        Ok(_) => {}
+                        Err(err) => {
+                            return Err(format!("Failed to read stdin: {}", err));
+                        }
+                    };
+                    in_line
+                },
+                Input::Source(reader) => {
+                    let mut in_line = String::new();
+                    let res = reader.read_line(&mut in_line);
+                    match res {
+                        Ok(0) => {
+                            if sourcing {
+                                input = original_input;
+                                original_input = Input::TempStore;
+                                sourcing = false;
+                                continue;
+                            }
+                            else {
+                                return Ok(());
+                            }
+                        }
+                        Ok(_) => {}
+                        Err(err) => {
+                            return Err(format!("Failed to read source file: {}", err));
+                        }
+                    };
+                    in_line
+                },
+                Input::TempStore => {
+                    panic!("Input got into illegal state");
+                }
+            };
+
+            // Check that input command ends with a semicolon
+            match line.find(';') {
+                Some(_) => {},
+                None => {
+                    if line.len() > 1 {
+                        eprintln!("Every command should end with ';': {}", line);
                     }
-                    Err(ReadlineError::Eof) => {
-                        println!("CTRL-D");
-                        save_history(&rl);
-                        return Ok(());
+                    
+                    continue;
+                },
+            };
+
+            let mut cmd_tokens = line.split_whitespace();
+            let first_cmd = match cmd_tokens.next() {
+                Some(cmd) => cmd.replace(";", ""),
+                None => { continue; },
+            };
+
+            match first_cmd.as_str() {
+                // TODO: what to do if source fails? partially executed OK, or do we want complete rollback?
+                "source" => {
+                    let filename = match cmd_tokens.next() {
+                        Some(filename) => filename.replace(";", ""),
+                        None => {
+                            eprintln!("Please provide source filename");
+                            eprintln!("Usage: source <filename>");
+                            continue;
+                        },
+                    };
+                    
+                    if let Ok(f) = File::open(filename) {
+                        sourcing = true;
+                        original_input = input;
```

Comment:
I know this is work in progress, but I guess the plan is to maintain a stack of inputs?  How is that related to the namespace stack?